docs/EVE-IMAGE-REPRODUCIBILITY.md

```diff
@@ -0,0 +1,274 @@
+* Regular automated audits of all `pkg/` sources.
+* CI scans of `pkg/` sources.
+
+## Linux Packages
+
+The primary source for many software packages in Linux is the OS packaging system. All of the OCI images used in EVE are based on
+[Alpine](https://hub.docker.com/_/alpine), which uses the [apk package manager](https://wiki.alpinelinux.org/wiki/Alpine_Package_Keeper).
+
+For example, to install bash:
+
+```sh
+apk add bash
+```
+
+Alpine packages creates unique challenges for reproducibility. Specifically:
+
+1. The version of the package changes when changing the alpine version. `apk add bash` can give entirely different bits when running on `alpine:3.15` vs `alpine:3.16`.
+1. The version of the package can change even _without_ changing the alpine version. `apk add bash` can give a different version running today on `alpine:3.16` vs yesterday or tomorrow on the same base `alpine:3.16`.
+
+It is possible to mitigate, but not completely solve, the above issue by adding a specific version:
+
+```sh
+apk add bash=5.1.16-r2
+```
+
+The above does not completely solve it, as sometimes alpine packages change their bits _even for the same version_. Thus, running
+`apk add bash=5.1.16-r2` on the same base `alpine:3.16` can give different bits on different days. Alpine packages are changed "under the covers"
+when bugs are found or updates are provided.
+
+In order to resolve this issue, both the external/upstream `linuxkit/*` packages and the `lfedge/eve-*` packages in this repository follow the same principles.
+
+1. Create an alpine "base" image: `linuxkit/alpine` and `lfedge/eve-alpine`
+1. All packages included in the composed EVE image **must** use `FROM lfedge/eve-alpine` or, in the case of linuxkit packages, `FROM linuxkit/alpine`, in their `Dockerfile`. `FROM alpine` is strictly forbidden.
+1. The base images source from upstream alpine, download all necessary apk packages needed by downstream, `lfedge/eve-alpine` for EVE packages and `linuxkit/alpine` for linuxkit packages, and cache them.
+1. When downstream images need to install packages, they run `apk add package` which sources it solely from the cache in the base image.
+
+The base images - `lfedge/eve-alpine` and `linuxkit/alpine` - act as pinning cache, with precise versions and bits of each package.
+
+Only when the cache is regenerated, i.e. when `lfedge/eve-alpine` or `linuxkit/alpine` is rebuilt, can these change.
+
+Further, the caches maintain hashes of the local cache contents, to ensure that any changes are known and captured.
+
+To make this more concrete, we will walk through a specific example, `lfedge/eve-storage-init`, which is based in [../pkg/storage-init](../pkg/storage-init/).
+
+The Dockerfile is as follows:
+
+```dockerfile
+FROM lfedge/eve-alpine:145f062a40639b6c65efa36bed1c5614b873be52 as build
+ENV PKGS alpine-baselayout musl-utils bash glib squashfs-tools util-linux e2fsprogs e2fsprogs-extra keyutils dosfstools coreutils sgdisk smartmontools
```

Comment:
storage-init does not have a separate BUILD_PKGS. Would it be better to use an example which does?

Reply:
changed to use `dom0-ztools` which has both, and added an explanation.